hermes-client/src/test/groovy/pl/allegro/tech/hermes/client/ReactiveHermesClientMicrometerMetricsTest.groovy

```diff
@@ -0,0 +1,132 @@
+import pl.allegro.tech.hermes.client.metrics.MetricsProvider
+import pl.allegro.tech.hermes.client.metrics.MicrometerMetricsProvider
+import reactor.core.publisher.Mono
+import spock.lang.Specification
+
+import java.time.Duration
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.TimeUnit
+
+import static pl.allegro.tech.hermes.client.ReactiveHermesClientBuilder.hermesClient
+
+class ReactiveHermesClientMicrometerMetricsTest extends Specification {
+
+    private final MeterRegistry metrics = new SimpleMeterRegistry()
+    private final MetricsProvider metricsProvider = new MicrometerMetricsProvider(metrics)
+
+    def "should measure publish latency"() {
+        given:
+        ReactiveHermesClient client = hermesClient(delayedHermesSender(Duration.ofMillis(100)))
+                .withRetrySleep(0)
+                .withMetrics(metricsProvider).build()
+
+        when:
+        client.publish("com.group.topic", "123").block()
+
+        then:
+        metrics.counter("hermes-client.com_group.topic.status.{code}", "code", String.valueOf(201)).count() == 1
+        def timer = metrics.timer("hermes-client.com_group.topic.latency")
+        timer.totalTime(TimeUnit.NANOSECONDS) >= Duration.ofMillis(100).get(ChronoUnit.NANOS)
+        timer.totalTime(TimeUnit.NANOSECONDS) < Duration.ofMillis(300).get(ChronoUnit.NANOS)
+    }
+
+    def "should close timer on exceptional completion and log failure metric"() {
+        given:
+        ReactiveHermesClient client = hermesClient({uri, msg ->  failingMono(new RuntimeException())})
+                .withRetrySleep(0)
+                .withRetries(3)
+                .withMetrics(metricsProvider).build()
+
+        when:
+        silence({ client.publish("com.group.topic", "123").block() })
+
+        then:
+        metrics.counter("hermes-client.com_group.topic.failure").count()  == 4
+        metrics.timer("hermes-client.com_group.topic.latency").count() == 4
+    }
+
+    def "should update max retries exceeded metric"() {
+        given:
+        ReactiveHermesClient client = hermesClient({uri, msg ->  failingMono(new RuntimeException())})
+                .withRetrySleep(0)
+                .withRetries(3)
+                .withMetrics(metricsProvider).build()
+
+        when:
+        silence({ client.publish("com.group.topic", "123").block() })
+
+        then:
+        metrics.counter("hermes-client.com_group.topic.failure").count() == 4
+        metrics.counter("hermes-client.com_group.topic.retries.count").count() == 3
+        metrics.counter("hermes-client.com_group.topic.retries.exhausted").count() == 1
+        metrics.counter("hermes-client.com_group.topic.retries.success").count() == 0
+        metrics.summary("hermes-client.com_group.topic.retries.attempts").takeSnapshot().count() == 0
+        metrics.timer("hermes-client.com_group.topic.latency").count() == 4
+    }
+
+    def "should update retries metrics"() {
```

Comment:
It would be nice to have more tests concerning metrics, with 0/1/2/3 out of 3 retries being performed and successful/failed.